File: EPI/10.11/swift/10.11.swift
public class Node<T> {
    public let value: T
    public let left: Node?
    public let right: Node?

    public init(value: T, left: Node<T>?, right: Node<T>?) {
        self.value = value
        self.left = left
        self.right = right
    }

    public convenience init?(preorder: [T?]) {
        if let root = reconstruct(preorder[0..<preorder.count]).root {
            self.init(value: root.value, left: root.left, right: root.right)
        } else {
            return nil
        }
    }
}

func reconstruct<T>(preorder: ArraySlice<T?>) -> (root: Node<T>?, count: Int) {
    if preorder.isEmpty || preorder.first! == nil {
        return (nil, 1)
    }

    let left = reconstruct(preorder[preorder.startIndex + 1..<preorder.endIndex])
    let right = reconstruct(preorder[preorder.startIndex + 1 + left.count..<preorder.endIndex])

    return (
        root: Node(value: preorder.first!!, left: left.root, right: right.root),
        count: 1 + left.count + right.count
    )
}

File: EPI/10.11/typescript/10.11.ts
export interface Node<T> {
    value: T;
    left?: this;
    right?: this;
}

export function constructFromPreorder<T>(preorder: (T | undefined)[]): Node<T> | undefined {
    return rec(preorder, 0).root;
}

function rec<T>(preorder: (T | undefined)[], i: number): {root: Node<T> | undefined, length: number} {
    const rootValue = preorder[i];

    if (rootValue === undefined) {
        return {root: undefined, length: 1};
    }

    const left = rec(preorder, i + 1);
    const right = rec(preorder, i + 1 + left.length);

    return {
        root: {value: rootValue, left: left.root, right: right.root},
        length: 1 + left.length + right.length
    };
}
